infer/resource.go

```diff
@@ -211,54 +214,127 @@ type fieldGenerator struct {
 	stateMatcher introspect.FieldMatcher
 	err          multierror.Error
 
+	fields map[string]*field
+}
+
+func (g *fieldGenerator) getField(name string) *field {
+	f, ok := g.fields[name]
+	if !ok {
+		f = new(field)
+		g.fields[name] = f
+	}
+	return f
+}
+
+type field struct {
 	// The set of tags that should always be secret
-	alwaysSecret map[string]bool
+	alwaysSecret bool
 	// The set of tags that should never be secret
-	neverSecret map[string]bool
-	// A map from a field to its dependencies
-	deps map[string][]string
+	neverSecret bool
+	// A map from a field to its dependencies.
+	deps []string
+
+	// If the output is known, regardless of other factors.
+	known bool
 }
 
 // MarkMap mutates m to comply with the result of the fieldGenerator, applying
 // computedness and secretness as appropriate.
-func (g *fieldGenerator) MarkMap(inputs, m resource.PropertyMap) {
-	// Flow secretness and computedness
-	for output, inputList := range g.deps {
-		output := resource.PropertyKey(output)
-		for _, input := range inputList {
-			input := inputs[resource.PropertyKey(input)]
-			if input.IsComputed() && !m[output].IsComputed() {
-				m[output] = resource.MakeComputed(m[output])
-				break
-			}
-			if input.IsSecret() && !m[output].IsSecret() {
-				m[output] = resource.MakeSecret(m[output])
-			}
+func (g *fieldGenerator) MarkMap(isCreate, isPreview bool) func(oldInputs, inputs, m resource.PropertyMap) {
+	return func(oldInputs, inputs, m resource.PropertyMap) {
+		// Flow secretness and computedness
+		for k, v := range m {
+			m[k] = markField(g.getField(string(k)), k, v, oldInputs, inputs, isCreate, isPreview)
 		}
 	}
+}
+
+func markComputed(
+	field *field, key resource.PropertyKey, prop resource.PropertyValue,
+	oldInputs, inputs resource.PropertyMap, isCreate bool,
+) resource.PropertyValue {
+	// If the value is already computed or if it is guaranteed to be known, we don't need to do anything
+	if field.known || prop.IsComputed() {
+		return prop
+	}
 
-	// Create mandatory secrets
-	for s := range g.alwaysSecret {
-		s := resource.PropertyKey(s)
-		if m[s].IsComputed() {
-			break
-		}
-		if !m[s].IsSecret() {
-			v := m[s]
-			m[s] = resource.NewSecretProperty(&resource.Secret{Element: v})
+	if input, ok := inputs[key]; ok && !input.IsComputed() && input.DeepEquals(prop) {
```

Comment:
OK we walked through this a little bit. The case @iwahbe made for marking things unknown when declared as such because the system is essentially slightly cheating /approximating propagation of unknowns. When passing data into the user code, unknowns are represented as zero values and the code is called anyway. When the output code received from the user code has zero values in unknown position it's' reasonable to assume these are unknowns. When the output code has non-zero values, they might have been a result of a computation like `x+y` that should have been `unknown+y=unknown` but actually works as `empty+y=y`, so that's the case for the framework discarding `y` and making things unknown when the static approximation specifies it should be unknown. 

The alternative we considered here is propagating this `y` out as known in spite of static specification to preserve more information.

Stepping back, this is heuristic land with no clear answers. What we'd like to have long term instead is an API for the user to hydrate PropertyValue to Go structures that's lossless with respect to unknown/empty distinction; basically if the provider author wants to handle unknowns then they bind to Go structures that can represent unknowns; and then if an incoming PropertyValue fails to bind to the Go structure then the framework either bypasses calling the user code all-together or errors out. We'll consider coming back to redesigning this pathway once there's more clarity on Go generics support in Pulumi..